File: src/Wails.Net.Errors/CallError.cs
namespace Wails.Net.Errors;

/// <summary>
/// 表示绑定调用失败的结构化错误。
/// 对应 Wails v3 Go 版本 bindings.go 中的 CallError 类型，
/// 用于将调用错误信息序列化为 JSON 并返回给前端。
/// </summary>
public class CallError
{
    /// <summary>
    /// 获取错误消息。
    /// </summary>
    public string Message { get; }

    /// <summary>
    /// 获取导致此错误的原始错误消息（可为空）。
    /// </summary>
    public string? Cause { get; }

    /// <summary>
    /// 获取错误类型。
    /// </summary>
    public CallErrorKind Kind { get; }

    /// <summary>
    /// 使用消息、原因和错误类型初始化 <see cref="CallError"/> 类的新实例。
    /// </summary>
    /// <param name="message">错误消息。</param>
    /// <param name="cause">导致此错误的原始错误消息。</param>
    /// <param name="kind">错误类型。</param>
    public CallError(string message, string? cause, CallErrorKind kind)
    {
        Message = message;
        Cause = cause;
        Kind = kind;
    }

    /// <summary>
    /// 返回包含所有字段的字符串表示形式。
    /// </summary>
    /// <returns>包含消息、原因和错误类型的字符串。</returns>
    public override string ToString()
    {
        var cause = Cause ?? "(none)";
        return $"Message: {Message}, Cause: {cause}, Kind: {Kind}";
    }

    /// <summary>
    /// 将此错误转换为可用于 JSON 序列化的字典。
    /// 字段名使用 camelCase 以符合前端 JSON 约定。
    /// </summary>
    /// <returns>包含 message、cause 和 kind 字段的字典。</returns>
    public Dictionary<string, object?> ToJson()
    {
        return new Dictionary<string, object?>
        {
            ["message"] = Message,
            ["cause"] = Cause,
            ["kind"] = Kind.ToString()
        };
    }
}

/// <summary>
/// 定义绑定调用错误的类型。
/// 对应 JavaScript 运行时的错误类型。
/// </summary>
public enum CallErrorKind
{
    /// <summary>
    /// 引用错误：访问未定义的变量或方法。
    /// </summary>
    ReferenceError = 0,

    /// <summary>
    /// 类型错误：参数类型不匹配或方法签名错误。
    /// </summary>
    TypeError = 1,

    /// <summary>
    /// 运行时错误：执行过程中发生的其他错误。
    /// </summary>
    RuntimeError = 2
}

File: packages/wails-net-runtime/src/internal/call-error.ts
import type { CallErrorKind, WailsCallError } from "./types.js";

/**
 * 后端返回的业务错误。继承自 `Error`，额外携带 `kind` 与 `cause`，
 * 便于调用方区分 `ReferenceError` / `TypeError` / `RuntimeError`（含权限拒绝）。
 */
export class CallError extends Error {
  /** 错误种类。 */
  public readonly kind: CallErrorKind;
  /** 底层原因（可能为 null）。 */
  public readonly callCause: string | null;

  constructor(error: WailsCallError) {
    super(error.message);
    this.name = "CallError";
    this.kind = error.kind;
    this.callCause = error.cause ?? null;
    // 保持原型链（ES5 继承兼容）。
    Object.setPrototypeOf(this, CallError.prototype);
  }
}

/**
 * 将后端返回的错误对象归一化为 `CallError`。
 * @param raw 可能为 `WailsCallError` 或字符串。
 */
export function toCallError(raw: WailsCallError | string | null | undefined): CallError {
  if (raw == null) {
    return new CallError({ message: "未知错误", cause: null, kind: "RuntimeError" });
  }
  if (typeof raw === "string") {
    return new CallError({ message: raw, cause: null, kind: "RuntimeError" });
  }
  return new CallError(raw);
}
